pkgs/stdenv/generic/check-meta.nix

```diff
@@ -250,58 +251,191 @@ let
+    };
+    insecure = lib.mkOption {
+      type = bool;
+      default = false;
+    };
     # TODO: refactor once something like Profpatsch's types-simple will land
     # This is currently dead code due to https://github.com/NixOS/nix/issues/2532
-    tests = attrsOf (mkOptionType {
-      name = "test";
-      check = x: x == {} || ( # Accept {} for tests that are unsupported
-        isDerivation x &&
-        x ? meta.timeout
-      );
-      merge = lib.options.mergeOneOption;
-    });
-    timeout = int;
+    tests = lib.mkOption {
+      type = attrsOf (mkOptionType {
+        name = "test";
+        check = x: x == {} || ( # Accept {} for tests that are unsupported
+          isDerivation x &&
+          x ? meta.timeout
+        );
+        merge = lib.options.mergeOneOption;
+      });
+      default = {};
+    };
+    timeout = lib.mkOption {
+      type = nullOr int;
+      default = null;
+    };
 
     # Weirder stuff that doesn't appear in the documentation?
-    maxSilent = int;
-    knownVulnerabilities = listOf str;
-    name = str;
-    version = str;
-    tag = str;
-    executables = listOf str;
-    outputsToInstall = listOf str;
-    position = str;
-    available = bool;
-    isBuildPythonPackage = platforms;
-    schedulingPriority = int;
-    isFcitxEngine = bool;
-    isIbusEngine = bool;
-    isGutenprint = bool;
-    badPlatforms = platforms;
+    maxSilent = lib.mkOption {
+      type = nullOr int;
+      default = null;
+    };
+    knownVulnerabilities = lib.mkOption {
+      type = listOf str;
+      default = [];
+    };
+    name = lib.mkOption {
+      type = nullOr str;
+      default = null;
+    };
+    version = lib.mkOption {
+      type = nullOr str;
+      default = null;
+    };
+    tag = lib.mkOption {
+      type = nullOr str;
+      default = null;
+    };
+    executables = lib.mkOption {
+      type = listOf str;
+      default = [];
+    };
+    outputsToInstall = lib.mkOption {
+      type = listOf str;
+      default = [];
+    };
+    position = lib.mkOption {
+      type = nullOr str;
+      default = null;
+    };
+    available = lib.mkOption {
+      type = bool;
+      default = false;
+    };
+    isBuildPythonPackage = lib.mkOption {
+      type = platforms.type;
+      default = [];
+    };
+    schedulingPriority = lib.mkOption {
+      type = nullOr int;
+      default = null;
+    };
+    isFcitxEngine = lib.mkOption {
+      type = bool;
+      default = false;
+    };
+    isIbusEngine = lib.mkOption {
+      type = bool;
+      default = false;
+    };
+    isGutenprint = lib.mkOption {
+      type = bool;
+      default = false;
+    };
+    badPlatforms = lib.mkOption {
+      type = platforms.type;
+      default = [];
+    };
   };
 
-  checkMetaAttr = k: v:
-    if metaTypes?${k} then
-      if metaTypes.${k}.check v then null else "key '${k}' has a value ${toString v} of an invalid type ${builtins.typeOf v}; expected ${metaTypes.${k}.description}"
```

Comment:
`.check` was wrong anyway, it doesn’t recurse into nested types.

Reply:
Only its name, `check`, is wrong. We need a function that evaluates as little as possible (WHNF) when checking definitions in the module system.